src/rpp/rpp/operators/take_until.hpp

```diff
@@ -0,0 +1,132 @@
+//                    TC Wang 2022 - present.
+// Distributed under the Boost Software License, Version 1.0.
+//    (See accompanying file LICENSE_1_0.txt or copy at
+//          https://www.boost.org/LICENSE_1_0.txt)
+//
+// Project home: https://github.com/victimsnino/ReactivePlusPlus
+//
+
+#pragma once
+
+#include <rpp/defs.hpp>
+#include <rpp/operators/fwd/take_until.hpp>
+#include <rpp/operators/first.hpp>
+#include <rpp/sources/create.hpp>
+#include <rpp/subscribers/constraints.hpp>
+#include <rpp/utils/functors.hpp>
+
+
+#include <rpp/operators/details/subscriber_with_state.hpp> // create_subscriber_with_state
+
+
+IMPLEMENTATION_FILE(take_until_tag);
+
+namespace rpp::details
+{
+
+/**
+ * Functor (type-erasure) of "take_until" for on_next operator.
+ */
+template<constraint::decayed_type Type>
+struct take_until_on_next
+{
+    void operator()(auto&& value,
+                    const auto& subscriber,
+                    const std::shared_ptr<first_state>& state) const
+    {
+        if (state->count > 0)
+            subscriber.on_next(std::forward<decltype(value)>(value));
+        else
+            subscriber.on_completed();
+    }
+};
+
+template<constraint::decayed_type Type>
+struct take_until_throttler
+{
+
+    /**
+     * Functor (type-erasure) of "take_until" for on_next operator.
+     */
+    void operator()(auto&&,
+                    const auto& subscriber,
+                    const std::shared_ptr<first_state>& state) const
+    {
+        if (state->count > 0)
+            --state->count;
+
+        if (state->count == 0)
+            subscriber.on_completed();
+
+//        decrement_notification_count(subscriber, state);
+    }
+
+//    /**
+//     * Functor (type-erasure) of "take_until" for on_error operator.
+//     */
+//    void operator()(const std::exception_ptr&,
+//                    const auto& subscriber,
+//                    const std::shared_ptr<first_state>& state) const
+//    {
+//        decrement_notification_count(subscriber, state);
+//    }
+//
+//    /**
+//     * Functor (type-erasure) of "take_until" for on_completed operator.
+//     */
+//    void operator()(const auto& subscriber,
+//                    const std::shared_ptr<first_state>& state) const
+//    {
+//        decrement_notification_count(subscriber, state);
+//    }
+//
+//    void decrement_notification_count(const auto& subscriber,
+//                                      const std::shared_ptr<first_state>& state) const
+//    {
+//        if (state->count > 0)
+//            --state->count;
+//
+//        if (state->count == 0)
+//            subscriber.on_completed();
+//    }
+};
+
+/**
+ * \brief "combine_latest" operator (an OperatorFn used by "lift").
+ */
+template<constraint::decayed_type Type, constraint::observable TTriggerObservable>
+struct take_until_impl
+{
+    using TriggerType = utils::extract_observable_type_t<TTriggerObservable>;
+
+    const TTriggerObservable m_until_observable;
```

Comment:
q: You mean implicit `m_until_observable =` using move assignment operator in constructor?

q: Why is it bad for disabling "move" constructor here? Is it safer that we disable "move" constructor?

Reply:
No, i mean remove `const` qualifier.

>  Is it safer that we disable "move" constructor?

In this case- no. It is state of `_impl` - so, it is state of observable, not subscriber. As a result, any observable with `take_until` on board would not be able to be moved, but it is bad. For example
```cpp
  obs.map(...)
     .take_untile(...) // starting from this one no any moves would be possible, only copies
     .filter(...)
     .take(...)
```

Reply:
Oh this is very interesting, I didn't know the observable may be copied in a chain. I thought the observable is always moved in order to save CPU time and memory. 

How do I know what operator copies the observable and what moves the observable?

Reply:
Also, if `take_until` moves the observable during the subscribing, the subsequent subscribing (could be triggered by `retry` family of operators) would get a segmentation fault?

Reply:
No, you didn't get my point.

When we call something like `my_obs.map(...)` we actually call `map` function with `const &` qualifier. For example
```cpp
template<std::invocable<Type> Callable>
    auto map(Callable&& callable) const & requires is_header_included<map_tag, Callable>
    {
        return static_cast<const SpecificObservable*>(this)->template lift<utils::decayed_invoke_result_t<Callable, Type>>(map_impl<Type, std::decay_t<Callable>>{std::forward<Callable>(callable)});
    }

```

What do we do there? Actually, we call `lift` from lvalue observable. Due to we construct new observable, we need to copy current to extend chain. So, observable is copied, because we need to keep both (`my_obs` and new one) as valid. Do you get it?

Then, we also can call same `map` as rvalue in both cases
1) `std::move(my_obs).map(...)` - manually move it
2) `my_obs.take(1).map(...)` - results of `take` is new observable, but currently is `prvalue`

In this case another overloading would be called:

```cpp
template<std::invocable<Type> Callable>
    auto map(Callable&& callable) && requires is_header_included<map_tag, Callable>
    {
        return std::move(*static_cast<SpecificObservable*>(this)).template lift<utils::decayed_invoke_result_t<Callable, Type>>(map_impl<Type, std::decay_t<Callable>>{std::forward<Callable>(callable)});
    }
```

It is rvalue overloading, so, we call lift as rvalue (via move). Same as above: to construct new observable we need have "old" observable to extend it. But due to we call it as rvalue we can "take" observable as is without copy. So, we just move current observable.

Observable is just "on subscribe" function. If "on subscribe" is not movable, then any move would be actually copy

Reply:
So, these copies/moves happen at the stage of construction of the chain, not subscribing